apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PgNumeric.cs

```diff
@@ -0,0 +1,481 @@
+        public override int GetHashCode() => GaxEqualityHelpers.CombineHashCodes(_notANumber ? -486187739 : -16777619, _scaledValue.GetHashCode(), _scale.GetHashCode());
+
+        /// <summary>
+        /// Parses a textual representation as a <see cref="PgNumeric"/>.
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        /// <paramref name="text"/> must be a representation of a decimal value which can be represented by <see cref="PgNumeric"/>,
+        /// using "." as a decimal place where one is specified, and a leading "-" for negative values. Leading zeroes and insignificant
+        /// trailing digits are permitted.
+        /// </para>
+        /// </remarks>
+        /// <param name="text">The text to parse. Must not be null.</param>
+        /// <returns>The parsed value.</returns>
+        /// <exception cref="FormatException">The value could not be parsed as a <see cref="PgNumeric"/>.</exception>
+        public static PgNumeric Parse(string text)
+        {
+            string message = TryParseImpl(text, out var value);
+            return message == null ? value : throw new FormatException(message);
+        }
+
+        /// <summary>
+        /// Attempts to parse a textual representation of a <see cref="PgNumeric"/> value.
+        /// </summary>
+        /// <remarks>
+        /// See <see cref="Parse"/> for format details. This method will return <c>true</c> if and only if
+        /// <see cref="Parse"/> would return without an exception.
+        /// </remarks>
+        /// <param name="text">The text to parse. Must not be null.</param>
+        /// <param name="value">The parsed value, or 0 on failure.</param>
+        /// <returns><c>true</c> if <paramref name="text"/> was parsed successfully; <c>false</c> otherwise.</returns>
+        public static bool TryParse(string text, out PgNumeric value)
+        {
+            string message = TryParseImpl(text, out value);
+            return message == null;
+        }
+
+        /// <summary>
+        /// Returns the error message for a FormatException if parsing fails, or null for success.
+        /// </summary>
+        private static string TryParseImpl(string text, out PgNumeric value)
+        {
+            GaxPreconditions.CheckNotNull(text, nameof(text));
+
+            if (string.Equals(text, NotANumber, StringComparison.InvariantCultureIgnoreCase))
+            {
+                value = new PgNumeric(default, 0, 0, true);
+                return null;
+            }
+
+            if (!s_validation.IsMatch(text))
+            {
+                value = default;
+                return "Text representation must be digits, containing an optional decimal point, and an optional leading '-' sign.";
+            }
+
+            int pointIndex = text.IndexOf('.');
+
+            // The raw value which will end up in the result, on success.
+            BigInteger rawValue;
+            int scale = 0;
```

Comment:
I'd declare this without a value, and assign a value of 0 in the `else` block, where you assign `rawValue`. That way it's clearer that the scale is zero when there's no decimal point.